extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+\valuemap{} or \tokenmap{} contains anything that needs to be processed.  If this is
+a concern then one could implement a model with two types of
+transaction, essentially just the disjoint union of the XUTXO-1 and
+XUTXO-2 or XUTXO-3 transaction types. A simple case distinction at the
+start of a transaction could then select either a fast
+native-currency-only computation or a slower multicurrency computation.
+This would be harder to maintain though.
+
+\medskip
+A more significant cost may be that we can no longer use \verb|{-# UNPACK #-}|
+when our Value type stops being a simple combination of wrappers
+and products around primitives, but this is again an issue with any
+multi-currency proposal.
+
+  
+\subsection{Further generalisations}
+We have assumed that our basic \valuetype{} type is the type of
+natural numbers.  However, the models mostly carry over to the case
+where \valuetype{} is an arbitrary abelian group\footnotemark.  We can
+even replace \valuetype{} with a commutative monoid as long as we
+relinquish the ability to destroy quantities of currencies (since we
+can no longer subtract in general). Whether these generalisations are
+useful in practice remains to be seen.
+
+\footnotetext{If we allow a non-abelian group then we may be in
+  trouble.  If we forge $a$ units of currency, then forge $b$ units,
+  then destroy $a$ units again, we're left with $aba^{-1}$ (or
+  $a+b-a$) units, which won't be equal to $b$ in general: if your
+  currency consists of elements of the free group on $\{a,b\}$ then
+  things could get pretty interesting.  Similar things may happen if
+  we need to split up an output to get some change. In the monoid
+  case, we might get unexpected cancellation as well: $a+b=0$, but
+  $a,b \ne 0$: consider $\mathbb{Z}_{24}$ or something. I suppose
+  doing things like this would let us write weird papers to impress
+  the cryptocurrency community. It could still be useful for something
+  though...}
+
+
+
+\section{What else do we need to worry about?}
+\begin{itemize}
+
+\item It might be worth adding an appendix that contains the entire
+  description of XUTXO-3 plus the relevant types and so on in one place.
+  
+\item Should we mention the problem of \textbf{validating the next data script}?
+%% https://github.com/input-output-hk/plutus/issues/426
+
+\item Sometimes I'm saying ``datascript'' and at other times ``data script''.
+  Maybe we should try to be consistent about this.
+  
+\item Do we need to worry about who signs the various scripts, or is that
+  something the wallet deals with?
+
+\item Who's responsible for looking after the scripts, and where are they kept?
+  Presumably scripts will be stored in the chain, but how do they get there?
+
+\item Is the number of inputs and/or outputs to a transaction always nonzero?
+  Now that we're allowed to destroy currency, maybe we can have transactions
+  with no outputs.
+
+\item Do we want to include \s{PayToPubKey} transactions in the
```

Comment:
Bruno's multicurrency paper ignores pay to pubkey transaction because they can be implemented as scripts. This makes sense as long as we don't have a cost model.

(Fees have to be paid by pay-to-pubkey transactions)

Reply:
Good point about the cost model.